coffee/session.coffee

```diff
@@ -5,70 +5,87 @@ palava = @palava
 
 # Session is a wrapper around a concrete room, channel and userMedia
 class palava.Session extends @EventEmitter
-
-  # @param o [Object] Options for the session
-  # @option o channel [palava.Channel] A channel Object which will be used for communication
-  # @option o web_socket_channel [WebSocket] A websocket from which a channel will be created
-  # @option o identity [palava.Identity] TODO
-  # @option o options [Object] TODO
-  #
   constructor: (o) ->
-    @resetOptions(o, false)
+    @roomOptions = {}
+    @assignOptions(o)
 
-  # Reset all options to default parameters
-  #
-  # @param o          [Object]  See constructor for details
-  # @param reconnect  [Boolean] Is part of a reconnect?
+  # Initializes the session and joins room when stream ready
   #
-  resetOptions: (o, reconnect = false) =>
-    @channel     = null
-    @userMedia   = null unless reconnect
-    @roomId      = null
-    @roomOptions = {}
+  # @param o [Object] Options for the session
+  # @option o webSocketAddress [WebSocket] The websocket endpoint to connect to
+  # @option o userMediaConfig [Object] getUserMedia constraints
+  # @option o stun [String] Address of stun server
+  # @option o turn [String] Turn address and credentials
+  # @option o joinTimeout [Integer] Milliseconds till joining is canceled by throwing
+  #           the "join_error" event
+  #
+  connect: (o) =>
     @assignOptions(o)
-    @initOptions = o
+    return unless @checkRequirements()
 
-  # Initializes the session
+    @createChannel()
+    @createRoom()
+
+    if @userMedia.stream
+      @room.join()
+    else
+      @userMedia.requestStream().then =>
+        @room.join()
+
+  # Reconnect the session
   #
-  # @param o [Object] See constructor for details
-  # @param reconnect  [Boolean] Is part of a reconnect?
+  reconnect: =>
+    @emit 'session_reconnect'
+    @clearConnection()
+    @createChannel()
+    @createRoom()
+    @room.join()
+
+  # Reset channel and room
   #
-  init: (o, reconnect = false) =>
-    @assignOptions(o)
-    Object.assign(@initOptions, o)
-    return unless @checkRequirements() unless reconnect
-    @setupRoom()
-    @userMedia.requestStream() unless reconnect
+  clearConnection: =>
+    @room.leave()
```

Comment:
`@room?.leave()`? (e.g. if somebody calls clearConnection twice for some reason)